include/zephyr/usb/usbd.h

```diff
@@ -0,0 +1,717 @@
+};
+
+/* TODO: Kconfig option USBD_NUMOF_INTERFACES_MAX? */
+#define USBD_NUMOF_INTERFACES_MAX	16U
+
+/**
+ * USB device support middle layer runtime state
+ *
+ * Part of USB device states without suspended and powered
+ * states, as it is better to track them separately.
+ */
+enum usbd_ch9_state {
+	USBD_STATE_DEFAULT = 0,
+	USBD_STATE_ADDRESS,
+	USBD_STATE_CONFIGURED,
+};
+
+
+/**
+ * USB device control request completion structure
+ */
+struct usbd_req_compl {
+	/** Setup request completion callback */
+	int (*cb)(void *const context, void *const data);
+	/** Setup request completion callback data */
+	void *data;
+};
+
+/**
+ * USB device support middle layer runtime data
+ */
+struct usbd_ch9_data {
+	/** Setup packet, up-to-date for the respective control request */
+	struct usb_setup_packet setup;
+	/** Control type, internaly used for stage verification */
+	int ctrl_type;
+	/** Protocol state of the USB device stack */
+	enum usbd_ch9_state state;
+	/** Halted endpoints bitmap */
+	uint32_t ep_halt;
+	/** USB device stack selected configuration */
+	uint8_t configuration;
+	/** Array to track interfaces alternate settings */
+	uint8_t alternate[USBD_NUMOF_INTERFACES_MAX];
+	/** Request completion structure */
+	struct usbd_req_compl req_compl;
+};
+
+/**
+ * USB device support status
+ */
+struct usbd_status {
+	/** USB device support is initialized */
+	unsigned int initialized : 1;
+	/** USB device support is enabled */
+	unsigned int enabled : 1;
+	/** USB device is suspended */
+	unsigned int suspended : 1;
+	/** USB remote wake-up feature is enabled */
+	unsigned int rwup : 1;
+};
+
+/**
+ * USB device support runtime context
+ *
+ * Main structure that organizes all descriptors, configuration,
+ * and interfaces. An UDC device must be assigned to this structure.
+ */
+struct usbd_contex {
+	/** Name of the USB device */
+	const char *name;
+	/** Access mutex */
+	struct k_mutex mutex;
+	/** Pointer to UDC device */
+	const struct device *dev;
+	/** Middle layer runtime data */
+	struct usbd_ch9_data ch9_data;
+	/** slist to manage descriptors like string, bos */
+	sys_slist_t descriptors;
+	/** slist to manage device configurations */
+	sys_slist_t configs;
+	/** Status of the USB device support */
+	struct usbd_status status;
+	/** Pointer to device descriptor */
+	void *desc;
+};
+
+/**
+ * @brief Vendor Requests Table
+ */
+struct usbd_cctx_vendor_req {
+	/** Array of vendor requests supportd by the class */
+	const uint8_t *reqs;
+	/** Length of the array */
+	uint8_t len;
+};
+
+/** USB Class instance registered flag */
+#define USBD_CCTX_REGISTERED		0
+
+struct usbd_class_node;
+
+/**
+ * @brief USB device support class instance API
+ */
+struct usbd_class_api {
+	/** Configuration update handler */
+	void (*update)(struct usbd_class_node *const node,
+		       uint8_t iface, uint8_t alternate);
+
+	/** USB control request handler to device */
+	int (*control_to_dev)(struct usbd_class_node *const node,
```

Comment:
Separating the to device and to host might seem good, but does not really comply to Universal Serial Bus Specification Revision 2.0, 9.3.1 bmRequestType
> This bitmapped field identifies the characteristics of the specific request. In particular, this field identifies the direction of data transfer in the second phase of the control transfer. The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

The device has to ignore the Direction bit when wLength is 0. Having separate handlers effectively prohibit the device from ignoring Direction bit.